notarization_poller/src/notarization_poller/worker.py

```diff
@@ -0,0 +1,152 @@
+
+    async def invoke(self):
+        """Claims and processes Taskcluster work."""
+        while not self.is_cancelled and not self.is_stopped:
+            num_tasks_to_claim = min(self.config["max_concurrent_tasks"] - len(self.running_tasks), MAX_CLAIM_WORK_TASKS)
+            if num_tasks_to_claim > 0 and arrow.utcnow().timestamp - self.last_claim_work.timestamp >= self.config["claim_work_interval"]:
+                async with aiohttp.ClientSession() as session:
+                    queue = Queue(
+                        options={
+                            "credentials": {"accessToken": self.config["taskcluster_access_token"], "clientId": self.config["taskcluster_client_id"]},
+                            "rootUrl": self.config["taskcluster_root_url"],
+                        },
+                        session=session,
+                    )
+                    new_tasks = await self._run_cancellable(claim_work(self.config, queue, num_tasks=num_tasks_to_claim)) or {}
+                self.last_claim_work = arrow.utcnow()
+                for claim_task in new_tasks.get("tasks", []):
+                    new_task = Task(self.config, claim_task)
+                    new_task.start()
+                    self.running_tasks.append(new_task)
+            await self.prune_running_tasks()
+            sleep_time = self.last_claim_work.timestamp + self.config["claim_work_interval"] - arrow.utcnow().timestamp
+            sleep_time > 0 and await self._run_cancellable(sleep(sleep_time))
+        self.running_tasks and await asyncio.wait([task.main_fut for task in self.running_tasks if task.main_fut])
+
+    async def prune_running_tasks(self):
+        """Prune any complete tasks from ``self.running_tasks``."""
+        for task in self.running_tasks:
+            if task.complete:
+                self.running_tasks.remove(task)
+
+    async def _run_cancellable(self, coroutine: typing.Awaitable):
+        if not self.is_cancelled:
+            self.future = asyncio.ensure_future(coroutine)
+            result = await self.future
+            self.future = None
+            return result
+
+    async def cancel(self, status=STATUSES["worker-shutdown"]):
+        """Cancel any running tasks."""
+        self.is_cancelled = True
+        self.future and self.future.cancel()
+        try:
+            for task in self.running_tasks:
+                task.task_fut and task.task_fut.cancel()
+            await asyncio.wait([task.main_fut for task in self.running_tasks if task.main_fut])
+        except (asyncio.CancelledError, ValueError):
+            pass
+
+
+# main {{{1
+def main(event_loop=None):
+    """Notarization poller entry point: get everything set up, then enter the main loop.
+
+    Args:
+        event_loop (asyncio.BaseEventLoop, optional): the event loop to use.
+            If None, use ``asyncio.get_event_loop()``. Defaults to None.
+
+    """
+    event_loop = event_loop or asyncio.get_event_loop()
+    done = False
```

Comment:
`done` appears to be set an never looked at, so can be removed.

Reply:
Done